video/out/opengl/context_cocoa.c

```diff
@@ -67,22 +67,31 @@ static CGLError test_gl_version(struct MPGLContext *ctx, CGLOpenGLProfile ver)
     struct priv *p = ctx->priv;
 
     CGLPixelFormatAttribute attrs[] = {
+        // let this array ordered by the inverse order of the most probably
+        // rejected attribute to preserve the fallback code
         kCGLPFAOpenGLProfile,
         (CGLPixelFormatAttribute) ver,
         kCGLPFAAccelerated,
-        // leave this as the last entry of the array to not break the fallback
-        // code
+        kCGLPFAAllowOfflineRenderers,
+        // keep this one last to apply the cocoa-force-dedicated-gpu option
         kCGLPFASupportsAutomaticGraphicsSwitching,
         0
     };
 
     GLint npix;
     CGLError err;
+    int supported_attribute = MP_ARRAY_SIZE(attrs)-1;
+
+    if (p->opts->cocoa_force_dedicated_gpu)
+        attrs[--supported_attribute] = 0;
+
     err = CGLChoosePixelFormat(attrs, &p->pix, &npix);
-    if (p->opts->cocoa_force_dedicated_gpu || err == kCGLBadAttribute) {
-        // kCGLPFASupportsAutomaticGraphicsSwitching is probably not supported
-        // by the current hardware. Falling back to not using it.
-        attrs[MP_ARRAY_SIZE(attrs) - 2] = 0;
+    while (err == kCGLBadAttribute) {
```

Comment:
this could potential lead to an endless loop or accessing something out of bounds, we need an explicit ending condition.

case of the endless loop/out of bounds is when the first opengl version (`kCGLOGLPVersion_3_2_Core`) is not supported. after removing kCGLPFAAllowOfflineRenderers and then kCGLPFAAccelerated, it would start removing the opengl version (probably leads to an invalid pixel format attribute array) till we have a completely empty array and then starting to access out of bounds of the array.

i would suggest a condition like `err == kCGLBadAttribute && supported_attribute > 3`. >3 so we won't touch the first 3 entries because we want to keep them.